{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a94d3b25",
   "metadata": {},
   "source": [
    "# Toy Example of traceCB"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "226f5829",
   "metadata": {},
   "source": [
    "In this notebook, we will use example data to run traceCB. First, load data from `data/toy_example/`."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2fe255f1",
   "metadata": {},
   "source": [
    "## Set up paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5c9197ae",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Main Directory set to: /home/wjiang49/traceCB\n",
      "Data Directory set to: /home/wjiang49/traceCB/data/toy_example\n"
     ]
    }
   ],
   "source": [
    "# set path and load data\n",
    "import os\n",
    "import pandas as pd\n",
    "\n",
    "main_dir = os.path.abspath(os.path.join(os.getcwd(), \"../..\"))\n",
    "print(\"Main Directory set to:\", main_dir)\n",
    "\n",
    "data_dir = os.path.join(main_dir, \"data\", \"toy_example\")\n",
    "print(\"Data Directory set to:\", data_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "847d69ec",
   "metadata": {},
   "source": [
    "Check whether the directory has been set up to: `<your_path>/traceCB`, and the data is in `<your_path>/traceCB/data/toy_example/`."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b6d24d2b",
   "metadata": {},
   "source": [
    "## Load and preprocess data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "27467f5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load data\n",
    "target_gene = \"ENSG00000025708\"  # Example gene\n",
    "## summary statistics for EAS, EUR, and tissue of EUR\n",
    "eas_summary_df = pd.read_csv(os.path.join(data_dir, \"eas_summary_statistics.csv\"))\n",
    "eur_summary_df = pd.read_csv(os.path.join(data_dir, \"eur_summary_statistics.csv\"))\n",
    "tissue_summary_df = pd.read_csv(os.path.join(data_dir, \"tissue_summary_statistics.csv\"))\n",
    "## ld matrix for EAS, EUR, and cross-population\n",
    "eas_ld_df = pd.read_csv(os.path.join(data_dir, \"eas_ld.csv\"))\n",
    "eur_ld_df = pd.read_csv(os.path.join(data_dir, \"eur_ld.csv\"))\n",
    "cross_ld_df = pd.read_csv(os.path.join(data_dir, \"cross_ld.csv\"))\n",
    "## cell type proportion in the tissue\n",
    "celltype_proportion_df = pd.read_csv(os.path.join(data_dir, \"celltype_proportion.csv\"))\n",
    "celltype_proportion_percentage = celltype_proportion_df.loc[\n",
    "    celltype_proportion_df[\"Cell_type\"] == \"Monocytes\", \"Proportion\"\n",
    "].iloc[0]\n",
    "celltype_proportion = celltype_proportion_percentage / 100.0\n",
    "## calculate a constant used in GMMtissue\n",
    "other_celltype_proportions = (\n",
    "    celltype_proportion_df.loc[\n",
    "        celltype_proportion_df[\"Cell_type\"] != \"Monocytes\", \"Proportion\"\n",
    "    ]\n",
    "    / 100.0\n",
    ").to_numpy()\n",
    "\n",
    "pii_pij_sum = sum(\n",
    "    [\n",
    "        i * j\n",
    "        for i in other_celltype_proportions\n",
    "        for j in other_celltype_proportions\n",
    "        if i != j\n",
    "    ]\n",
    ")\n",
    "pi2_omega_sum_const = 2 * celltype_proportion + pii_pij_sum / (1 - celltype_proportion)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "527e739c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Head of Summary DataFrames should look like this:\n",
      "          RSID             GENE      BETA        SE         Z      PVAL    N\n",
      "0  rs75674482  ENSG00000025708  0.270058  0.253239  1.066417  0.288728  105\n",
      "1    rs135845  ENSG00000025708  0.223593  0.141665  1.578320  0.117558  105\n",
      "2    rs135846  ENSG00000025708  0.224359  0.141582  1.584656  0.116110  105\n",
      "3   rs8141979  ENSG00000025708  0.156355  0.148230  1.054817  0.293977  105\n",
      "4  rs35089849  ENSG00000025708  0.248425  0.253238  0.980993  0.328895  105\n",
      "Head of LD DataFrames should look like this:\n",
      "    CHR          SNP        BP  ENSG00000025708\n",
      "0   22    rs2005631  16869565              0.0\n",
      "1   22   rs35416799  16869887              0.0\n",
      "2   22  rs188798391  16878941              0.0\n",
      "3   22  rs186070939  16879285              0.0\n",
      "4   22  rs141431439  16880117              0.0\n"
     ]
    }
   ],
   "source": [
    "print(\"Head of Summary DataFrames should look like this:\\n\", eas_summary_df.head())\n",
    "print(\"Head of LD DataFrames should look like this:\\n\", eas_ld_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "af9919f4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Head of aligned gene DataFrame should looks like:\n",
      "          RSID             GENE  BETA_tar    SE_tar     Z_tar  PVAL_tar  N_tar  \\\n",
      "0  rs75674482  ENSG00000025708  0.270058  0.253239  1.066417  0.288728    105   \n",
      "1    rs135845  ENSG00000025708  0.223593  0.141665  1.578320  0.117558    105   \n",
      "2    rs135846  ENSG00000025708  0.224359  0.141582  1.584656  0.116110    105   \n",
      "3   rs8141979  ENSG00000025708  0.156355  0.148230  1.054817  0.293977    105   \n",
      "4  rs35089849  ENSG00000025708  0.248425  0.253238  0.980993  0.328895    105   \n",
      "\n",
      "   BETA_aux    SE_aux     Z_aux  PVAL_aux  N_aux  BETA_tissue  SE_tissue  \\\n",
      "0 -0.127073  0.218847 -0.580648  0.562223    191     0.017989   0.045871   \n",
      "1 -0.457850  0.185965 -2.462022  0.014779    191     0.031253   0.030403   \n",
      "2 -0.451186  0.185522 -2.431981  0.016018    191     0.020958   0.028198   \n",
      "3 -0.151611  0.113418 -1.336745  0.183028    191     0.006633   0.021028   \n",
      "4  0.112325  0.148428  0.756764  0.450204    191     0.009180   0.031176   \n",
      "\n",
      "   Z_tissue  PVAL_tissue  N_tissue     LD_tar     LD_aux       LD_x  \n",
      "0  0.392175     0.695068       670  41.499378   7.778024  10.882746  \n",
      "1  1.027951     0.304388       670  78.486038  48.717865  -4.070536  \n",
      "2  0.743257     0.457617       670  78.486038  50.023804  -4.009035  \n",
      "3  0.315417     0.752555       670  68.888725  44.522377   6.384441  \n",
      "4  0.294442     0.768522       670  42.443920  36.674423  22.165127  \n"
     ]
    }
   ],
   "source": [
    "## TL;DR: Align all the input data by RSID/SNP\n",
    "gene_df = (\n",
    "    eas_summary_df.merge(\n",
    "        eur_summary_df.drop(columns=[\"GENE\"]),\n",
    "        on=\"RSID\",\n",
    "        suffixes=(\"_tar\", \"_aux\"),\n",
    "        how=\"inner\",\n",
    "    )\n",
    "    .merge(\n",
    "        tissue_summary_df.drop(columns=[\"GENE\"]).rename(\n",
    "            columns={\n",
    "                col: f\"{col}_tissue\"\n",
    "                for col in tissue_summary_df.columns\n",
    "                if col != \"RSID\"\n",
    "            }\n",
    "        ),\n",
    "        on=\"RSID\",\n",
    "    )\n",
    "    .merge(\n",
    "        eas_ld_df.rename(columns={target_gene: \"LD_tar\", \"SNP\": \"RSID\"}).loc[\n",
    "            :, [\"RSID\", \"LD_tar\"]\n",
    "        ],\n",
    "        on=\"RSID\",\n",
    "        how=\"inner\",\n",
    "    )\n",
    "    .merge(\n",
    "        eur_ld_df.rename(columns={target_gene: \"LD_aux\", \"SNP\": \"RSID\"}).loc[\n",
    "            :, [\"RSID\", \"LD_aux\"]\n",
    "        ],\n",
    "        on=\"RSID\",\n",
    "        how=\"inner\",\n",
    "    )\n",
    "    .merge(\n",
    "        cross_ld_df.rename(columns={target_gene: \"LD_x\", \"SNP\": \"RSID\"}).loc[\n",
    "            :, [\"RSID\", \"LD_x\"]\n",
    "        ],\n",
    "        on=\"RSID\",\n",
    "        how=\"inner\",\n",
    "    )\n",
    ")\n",
    "print(\"Head of aligned gene DataFrame should looks like:\\n\", gene_df.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa3033cb",
   "metadata": {},
   "source": [
    "## LDSC and GMM (traceC and traceCB)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "02f48c2c",
   "metadata": {},
   "source": [
    "Then, load necessary libraries from `src`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d937d818",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "\n",
    "# Add the parent directory to the system path to import modules from 'src'\n",
    "module_path = os.path.abspath(os.path.join(os.getcwd(), \"../src\"))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "\n",
    "import numpy as np\n",
    "from traceCB.ldsc import Run_Cross_LDSC\n",
    "from traceCB.gmm import GMM, GMMtissue\n",
    "from traceCB.run_gmm import clip_correlation\n",
    "from traceCB.utils import z2p, MIN_FLOAT"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e12e3659",
   "metadata": {},
   "source": [
    "### Step 1a: Calculate heritability and co-heritability using `Run_Cross_LDSC` (cross population LD score regression)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "40c64ec4",
   "metadata": {},
   "outputs": [],
   "source": [
    "Omega, Omega_se = Run_Cross_LDSC(\n",
    "    (gene_df.BETA_tar / gene_df.SE_tar).to_numpy(),\n",
    "    gene_df.N_tar.to_numpy(),\n",
    "    gene_df.LD_tar.to_numpy(),\n",
    "    (gene_df.BETA_aux / gene_df.SE_aux).to_numpy(),\n",
    "    gene_df.N_aux.to_numpy(),\n",
    "    gene_df.LD_aux.to_numpy(),\n",
    "    gene_df.LD_x.to_numpy(),\n",
    "    np.array([1, 1, 0]),  # fixed intercepts\n",
    ")\n",
    "Omega_p = z2p(Omega / Omega_se)\n",
    "### threshold Omega\n",
    "for i in range(Omega.shape[0]):\n",
    "    for j in range(Omega.shape[1]):\n",
    "        if Omega_p[i, j] > 0.10:\n",
    "            Omega[i, j] = MIN_FLOAT\n",
    "if np.any(Omega <= MIN_FLOAT):\n",
    "    print(\"Co-heritability not significant, please try another gene\")\n",
    "Omega[0, 1], cor_clipped = clip_correlation(Omega[0, 0], Omega[1, 1], Omega[0, 1])\n",
    "Omega[1, 0] = Omega[0, 1]\n",
    "Omega_p = z2p(Omega / Omega_se)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cbf04914",
   "metadata": {},
   "source": [
    "### Step 1b: Calculate $\\Sigma_O$ using Run_Cross_LDSC with European ct-eQTL and tissue eQTL."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c323f104",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate Sigma_o\n",
    "pi2_omega_sum = 0.0\n",
    "run_gmm_tissue = True  # set default to run GMM with tissue\n",
    "\n",
    "aux_Omega_matrix, aux_Omega_matrix_se = Run_Cross_LDSC(\n",
    "    gene_df.BETA_aux / gene_df.SE_aux,\n",
    "    gene_df.N_aux,\n",
    "    gene_df.LD_aux,\n",
    "    gene_df.BETA_tissue / gene_df.SE_tissue,\n",
    "    gene_df.N_tissue,\n",
    "    gene_df.LD_aux,\n",
    "    gene_df.LD_aux,\n",
    "    np.array([1.0, 1.0, 0.0]),\n",
    ")\n",
    "\n",
    "aux_Omega_matrix[0, 1], _ = clip_correlation(\n",
    "    aux_Omega_matrix[0, 0], aux_Omega_matrix[1, 1], aux_Omega_matrix[0, 1]\n",
    ")\n",
    "aux_Omega_matrix[1, 0] = aux_Omega_matrix[0, 1]\n",
    "\n",
    "if np.all(z2p(aux_Omega_matrix / aux_Omega_matrix_se) < 0.10):\n",
    "    pi2_omega_sum = (\n",
    "        aux_Omega_matrix[1, 1]\n",
    "        - celltype_proportion**2 * Omega[1, 1]\n",
    "        - pi2_omega_sum_const\n",
    "        * np.maximum(aux_Omega_matrix[0, 1] - celltype_proportion * Omega[1, 1], 0)\n",
    "    )\n",
    "    if pi2_omega_sum < 0:\n",
    "        run_gmm_tissue = False\n",
    "else:\n",
    "    run_gmm_tissue = False\n",
    "\n",
    "if not run_gmm_tissue:\n",
    "    print(\n",
    "        \"GMMtissue will not be run due to non-significant co-heritability with tissue or negative SIGMAO. Please try another gene.\"\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a0d9d872",
   "metadata": {},
   "source": [
    "Step 2: Run core `GMM` (TraceC) and `GMMTissue` (TraceCB) algorithm."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "9dec5290",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Head of GMM results DataFrame should look like:\n",
      "          RSID  TAR_SBETA  TAR_CBETA  TAR_TBETA   TAR_SSE   TAR_CSE   TAR_TSE  \\\n",
      "0  rs75674482   0.270058   0.185148   0.193276  0.253239  0.235911  0.230797   \n",
      "1    rs135845   0.223593   0.236687   0.235555  0.141665  0.141524  0.141511   \n",
      "2    rs135846   0.224359   0.237014   0.236139  0.141582  0.141446  0.141434   \n",
      "3   rs8141979   0.156355   0.136936   0.137520  0.148230  0.147208  0.147179   \n",
      "4  rs35089849   0.248425   0.205571   0.200733  0.253238  0.176277  0.173052   \n",
      "\n",
      "     TAR_SZ    TAR_CZ    TAR_TZ  TAR_SPVAL  TAR_CPVAL  TAR_TPVAL  AUX_SBETA  \\\n",
      "0  1.066417  0.784823  0.837427   0.288728   0.432557   0.402352  -0.127073   \n",
      "1  1.578320  1.672418  1.664561   0.117558   0.094442   0.096000  -0.457850   \n",
      "2  1.584656  1.675657  1.669608   0.116110   0.093805   0.094997  -0.451186   \n",
      "3  1.054817  0.930220  0.934374   0.293977   0.352257   0.350111  -0.151611   \n",
      "4  0.980993  1.166183  1.159954   0.328895   0.243541   0.246068   0.112325   \n",
      "\n",
      "   AUX_CBETA  AUX_TBETA   AUX_SSE   AUX_CSE   AUX_TSE    AUX_SZ    AUX_CZ  \\\n",
      "0   0.041645   0.054795  0.218847  0.163752  0.149740 -0.580648  0.254319   \n",
      "1  -0.471712  -0.400154  0.185965  0.185548  0.175567 -2.462022 -2.542262   \n",
      "2  -0.464592  -0.401491  0.185522  0.185141  0.175247 -2.431981 -2.509400   \n",
      "3  -0.143906  -0.134727  0.113418  0.113144  0.110361 -1.336745 -1.271883   \n",
      "4   0.139583   0.133318  0.148428  0.143685  0.137745  0.756764  0.971451   \n",
      "\n",
      "     AUX_TZ  AUX_SPVAL  AUX_CPVAL  AUX_TPVAL  BETA_tissue  SE_tissue  \\\n",
      "0  0.365936   0.562223   0.799249   0.714413     0.017989   0.045871   \n",
      "1 -2.279209   0.014779   0.011014   0.022655     0.031253   0.030403   \n",
      "2 -2.290995   0.016018   0.012094   0.021964     0.020958   0.028198   \n",
      "3 -1.220791   0.183028   0.203415   0.222165     0.006633   0.021028   \n",
      "4  0.967861   0.450204   0.331324   0.333114     0.009180   0.031176   \n",
      "\n",
      "   Z_tissue  PVAL_tissue  \n",
      "0  0.392175     0.695068  \n",
      "1  1.027951     0.304388  \n",
      "2  0.743257     0.457617  \n",
      "3  0.315417     0.752555  \n",
      "4  0.294442     0.768522  \n"
     ]
    }
   ],
   "source": [
    "# Define functions to apply GMM and GMMtissue to each row\n",
    "def apply_gmm(row):\n",
    "    return GMM(\n",
    "        Omega,\n",
    "        np.eye(2),\n",
    "        row[\"BETA_tar\"],\n",
    "        row[\"SE_tar\"],\n",
    "        row[\"LD_tar\"],\n",
    "        row[\"BETA_aux\"],\n",
    "        row[\"SE_aux\"],\n",
    "        row[\"LD_aux\"],\n",
    "        row[\"LD_x\"],\n",
    "    )\n",
    "\n",
    "\n",
    "def apply_gmm_tissue(row):\n",
    "    return GMMtissue(\n",
    "        Omega,\n",
    "        np.eye(3),\n",
    "        row[\"BETA_tar\"],\n",
    "        row[\"SE_tar\"],\n",
    "        row[\"LD_tar\"],\n",
    "        row[\"BETA_aux\"],\n",
    "        row[\"SE_aux\"],\n",
    "        row[\"LD_aux\"],\n",
    "        row[\"LD_x\"],\n",
    "        row[\"BETA_tissue\"],\n",
    "        row[\"SE_tissue\"],\n",
    "        pi2_omega_sum,\n",
    "        celltype_proportion,\n",
    "    )\n",
    "\n",
    "\n",
    "# Apply the functions to the DataFrame, or you can use other parallelization methods\n",
    "gmm_c_results = gene_df.apply(apply_gmm, axis=1, result_type=\"expand\")\n",
    "gmm_c_results.columns = [\"TAR_CBETA\", \"TAR_CSE\", \"AUX_CBETA\", \"AUX_CSE\"]\n",
    "\n",
    "gmm_t_results = gene_df.apply(apply_gmm_tissue, axis=1, result_type=\"expand\")\n",
    "gmm_t_results.columns = [\"TAR_TBETA\", \"TAR_TSE\", \"AUX_TBETA\", \"AUX_TSE\"]\n",
    "\n",
    "# Combine original and new results into a single DataFrame\n",
    "gmm_results = pd.concat(\n",
    "    [\n",
    "        gene_df[\n",
    "            [\n",
    "                \"RSID\",\n",
    "                \"BETA_tar\",\n",
    "                \"SE_tar\",\n",
    "                \"Z_tar\",\n",
    "                \"PVAL_tar\",\n",
    "                \"BETA_aux\",\n",
    "                \"SE_aux\",\n",
    "                \"Z_aux\",\n",
    "                \"PVAL_aux\",\n",
    "                \"BETA_tissue\",\n",
    "                \"SE_tissue\",\n",
    "                \"Z_tissue\",\n",
    "                \"PVAL_tissue\",\n",
    "            ]\n",
    "        ].copy(),\n",
    "        gmm_c_results,\n",
    "        gmm_t_results,\n",
    "    ],\n",
    "    axis=1,\n",
    ")\n",
    "\n",
    "# Rename columns to match the original format\n",
    "gmm_results.rename(\n",
    "    columns={\n",
    "        \"BETA_tar\": \"TAR_SBETA\",\n",
    "        \"SE_tar\": \"TAR_SSE\",\n",
    "        \"Z_tar\": \"TAR_SZ\",\n",
    "        \"PVAL_tar\": \"TAR_SPVAL\",\n",
    "        \"BETA_aux\": \"AUX_SBETA\",\n",
    "        \"SE_aux\": \"AUX_SSE\",\n",
    "        \"Z_aux\": \"AUX_SZ\",\n",
    "        \"PVAL_aux\": \"AUX_SPVAL\",\n",
    "    },\n",
    "    inplace=True,\n",
    ")\n",
    "\n",
    "# Calculate Z-scores and P-values for the new GMM results\n",
    "for model in [\"C\", \"T\"]:  # C for cross, T for tissue\n",
    "    for pop in [\"TAR\", \"AUX\"]:\n",
    "        beta_col = f\"{pop}_{model}BETA\"\n",
    "        se_col = f\"{pop}_{model}SE\"\n",
    "        z_col = f\"{pop}_{model}Z\"\n",
    "        pval_col = f\"{pop}_{model}PVAL\"\n",
    "\n",
    "        gmm_results.loc[:, z_col] = gmm_results[beta_col] / gmm_results[se_col]\n",
    "        gmm_results.loc[:, pval_col] = z2p(gmm_results[z_col])\n",
    "\n",
    "# Reorder columns\n",
    "final_columns = [\n",
    "    \"RSID\",\n",
    "    \"TAR_SBETA\",\n",
    "    \"TAR_CBETA\",\n",
    "    \"TAR_TBETA\",\n",
    "    \"TAR_SSE\",\n",
    "    \"TAR_CSE\",\n",
    "    \"TAR_TSE\",\n",
    "    \"TAR_SZ\",\n",
    "    \"TAR_CZ\",\n",
    "    \"TAR_TZ\",\n",
    "    \"TAR_SPVAL\",\n",
    "    \"TAR_CPVAL\",\n",
    "    \"TAR_TPVAL\",\n",
    "    \"AUX_SBETA\",\n",
    "    \"AUX_CBETA\",\n",
    "    \"AUX_TBETA\",\n",
    "    \"AUX_SSE\",\n",
    "    \"AUX_CSE\",\n",
    "    \"AUX_TSE\",\n",
    "    \"AUX_SZ\",\n",
    "    \"AUX_CZ\",\n",
    "    \"AUX_TZ\",\n",
    "    \"AUX_SPVAL\",\n",
    "    \"AUX_CPVAL\",\n",
    "    \"AUX_TPVAL\",\n",
    "    \"BETA_tissue\",\n",
    "    \"SE_tissue\",\n",
    "    \"Z_tissue\",\n",
    "    \"PVAL_tissue\",\n",
    "]\n",
    "gmm_results = gmm_results[final_columns]\n",
    "\n",
    "pd.set_option(\"display.max_columns\", 30)\n",
    "print(\"Head of GMM results DataFrame should look like:\\n\", gmm_results.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "6e90145a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The columns of the GMM results DataFrame are:\n",
      " ['RSID', 'TAR_SBETA', 'TAR_CBETA', 'TAR_TBETA', 'TAR_SSE', 'TAR_CSE', 'TAR_TSE', 'TAR_SZ', 'TAR_CZ', 'TAR_TZ', 'TAR_SPVAL', 'TAR_CPVAL', 'TAR_TPVAL', 'AUX_SBETA', 'AUX_CBETA', 'AUX_TBETA', 'AUX_SSE', 'AUX_CSE', 'AUX_TSE', 'AUX_SZ', 'AUX_CZ', 'AUX_TZ', 'AUX_SPVAL', 'AUX_CPVAL', 'AUX_TPVAL', 'BETA_tissue', 'SE_tissue', 'Z_tissue', 'PVAL_tissue']\n"
     ]
    }
   ],
   "source": [
    "print(\"The columns of the GMM results DataFrame are:\\n\", gmm_results.columns.tolist())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "603f9029",
   "metadata": {},
   "source": [
    "where 'TAR_\\*' means the target population (here is EAS), 'AUX_\\*' means the auxiliary population (here is EUR), and '\\*_S\\*', '\\*_C\\*', '\\*_T\\*' means the summary statistics, cross-population enhancement, and cross-population with tissue enhancement results, respectively.\n",
    "\n",
    "If you would like to calculate the effect sample size of each result, you can use the `calculate_Neff` function from `src.traceCB.run_gmm`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "c7d3e7cf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Effect sample size of:\n",
      "summary statistics: 68.1234527359333,\n",
      "traceC enhancement: 98.44082852444448,\n",
      "traceCB enhancement: 107.30914032841594\n"
     ]
    }
   ],
   "source": [
    "from traceCB.run_gmm import calculate_Neff\n",
    "\n",
    "## calculate neff of summary statistics in eas\n",
    "tar_sneff = calculate_Neff(\n",
    "    gmm_results.TAR_SZ,\n",
    "    Omega[0, 0],\n",
    "    gene_df.LD_tar,\n",
    ")\n",
    "## calculate neff of traceC enhancement in eas\n",
    "tar_cneff = calculate_Neff(\n",
    "    gmm_results.TAR_CZ,\n",
    "    Omega[0, 0],\n",
    "    gene_df.LD_tar,\n",
    ")\n",
    "## calculate neff of traceCB enhancement in eas\n",
    "tar_tneff = calculate_Neff(\n",
    "    gmm_results.TAR_TZ,\n",
    "    Omega[0, 0],\n",
    "    gene_df.LD_tar,\n",
    ")\n",
    "print(\n",
    "    f\"Effect sample size of:\\nsummary statistics: {tar_sneff},\\ntraceC enhancement: {tar_cneff},\\ntraceCB enhancement: {tar_tneff}\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e9670d3",
   "metadata": {},
   "source": [
    "That's it! You can now use traceCB to analyze your data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a12aeba",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "py312",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
