
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="Trans-ancestry cell-type-specific eQTL effects mapping">
      
      
        <meta name="author" content="Luca Jiang">
      
      
        <link rel="canonical" href="https://lucajiang.github.io/traceCB/pipeline/">
      
      
        <link rel="prev" href="../tutorial/">
      
      
        <link rel="next" href="../api_reference/">
      
      
        
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.7.2">
    
    
      
        <title>Pipeline - traceCB</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.484c7ddc.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.ab4e12ef.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#howto-run-tracecb-pipeline" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="traceCB" class="md-header__button md-logo" aria-label="traceCB" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            traceCB
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Pipeline
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="slate" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12s-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/lucajiang/traceCB" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    lucajiang/traceCB
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
            
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href=".." class="md-tabs__link">
        
  
  
    
  
  Home

      </a>
    </li>
  

      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../tutorial/" class="md-tabs__link">
        
  
  
    
  
  Tutorial

      </a>
    </li>
  

      
        
  
  
  
    
  
  
    <li class="md-tabs__item md-tabs__item--active">
      <a href="./" class="md-tabs__link">
        
  
  
    
  
  Pipeline

      </a>
    </li>
  

      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../api_reference/" class="md-tabs__link">
        
  
  
    
  
  API Reference

      </a>
    </li>
  

      
    </ul>
  </div>
</nav>
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="traceCB" class="md-nav__button md-logo" aria-label="traceCB" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    traceCB
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/lucajiang/traceCB" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    lucajiang/traceCB
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Home
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../tutorial/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Tutorial
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    
  
    Pipeline
  

    
  </span>
  
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    
  
    Pipeline
  

    
  </span>
  
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#workflow-overview" class="md-nav__link">
    <span class="md-ellipsis">
      
        Workflow Overview
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#preprocessing" class="md-nav__link">
    <span class="md-ellipsis">
      
        Preprocessing
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Preprocessing">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#format-data-by-chromosome" class="md-nav__link">
    <span class="md-ellipsis">
      
        Format Data by Chromosome
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#cell-type-information-proportion" class="md-nav__link">
    <span class="md-ellipsis">
      
        Cell Type Information &amp; Proportion
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Cell Type Information &amp; Proportion">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#eqtlcatalogue-id-and-cell-types" class="md-nav__link">
    <span class="md-ellipsis">
      
        eQTLCatalogue ID and Cell Types
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#proportion-calculation" class="md-nav__link">
    <span class="md-ellipsis">
      
        Proportion Calculation
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#alignment" class="md-nav__link">
    <span class="md-ellipsis">
      
        Alignment
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#annotation-ld-scores" class="md-nav__link">
    <span class="md-ellipsis">
      
        Annotation &amp; LD Scores
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Annotation &amp; LD Scores">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-annotate-ld" class="md-nav__link">
    <span class="md-ellipsis">
      
        1. Annotate LD
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-run-s-ldxr" class="md-nav__link">
    <span class="md-ellipsis">
      
        2. Run s-ldxr
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gmm-modeling" class="md-nav__link">
    <span class="md-ellipsis">
      
        GMM Modeling
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="GMM Modeling">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#output-files" class="md-nav__link">
    <span class="md-ellipsis">
      
        Output Files
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#colocalization" class="md-nav__link">
    <span class="md-ellipsis">
      
        Colocalization
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Colocalization">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#prerequisites" class="md-nav__link">
    <span class="md-ellipsis">
      
        Prerequisites
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#1-leading-snp-annotation" class="md-nav__link">
    <span class="md-ellipsis">
      
        1. Leading SNP Annotation
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-run-coloc" class="md-nav__link">
    <span class="md-ellipsis">
      
        2. Run COLOC
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#visualization" class="md-nav__link">
    <span class="md-ellipsis">
      
        Visualization
      
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../api_reference/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    API Reference
  

    
  </span>
  
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#workflow-overview" class="md-nav__link">
    <span class="md-ellipsis">
      
        Workflow Overview
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#preprocessing" class="md-nav__link">
    <span class="md-ellipsis">
      
        Preprocessing
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Preprocessing">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#format-data-by-chromosome" class="md-nav__link">
    <span class="md-ellipsis">
      
        Format Data by Chromosome
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#cell-type-information-proportion" class="md-nav__link">
    <span class="md-ellipsis">
      
        Cell Type Information &amp; Proportion
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Cell Type Information &amp; Proportion">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#eqtlcatalogue-id-and-cell-types" class="md-nav__link">
    <span class="md-ellipsis">
      
        eQTLCatalogue ID and Cell Types
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#proportion-calculation" class="md-nav__link">
    <span class="md-ellipsis">
      
        Proportion Calculation
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#alignment" class="md-nav__link">
    <span class="md-ellipsis">
      
        Alignment
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#annotation-ld-scores" class="md-nav__link">
    <span class="md-ellipsis">
      
        Annotation &amp; LD Scores
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Annotation &amp; LD Scores">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-annotate-ld" class="md-nav__link">
    <span class="md-ellipsis">
      
        1. Annotate LD
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-run-s-ldxr" class="md-nav__link">
    <span class="md-ellipsis">
      
        2. Run s-ldxr
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gmm-modeling" class="md-nav__link">
    <span class="md-ellipsis">
      
        GMM Modeling
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="GMM Modeling">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#output-files" class="md-nav__link">
    <span class="md-ellipsis">
      
        Output Files
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#colocalization" class="md-nav__link">
    <span class="md-ellipsis">
      
        Colocalization
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Colocalization">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#prerequisites" class="md-nav__link">
    <span class="md-ellipsis">
      
        Prerequisites
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#1-leading-snp-annotation" class="md-nav__link">
    <span class="md-ellipsis">
      
        1. Leading SNP Annotation
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-run-coloc" class="md-nav__link">
    <span class="md-ellipsis">
      
        2. Run COLOC
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#visualization" class="md-nav__link">
    <span class="md-ellipsis">
      
        Visualization
      
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              
              <article class="md-content__inner md-typeset">
                
                  


  
  


<h1 id="howto-run-tracecb-pipeline">HowTO: Run traceCB pipeline</h1>
<p>This guide details the steps to preprocess data, execute the traceCB GMM model, and visualize the results.</p>
<p>All logs will be saved to your specified <code>log_path</code>. Please review the log files carefully for any warnings or errors.</p>
<div class="admonition warning data format requirement">
<p class="admonition-title">Warning</p>
<p>If you use your own eQTL data, ensure the input data format matches the specifications described below exactly to avoid runtime errors.</p>
</div>
<h2 id="workflow-overview">Workflow Overview</h2>
<pre class="mermaid"><code>graph LR
    A[Raw Data] --&gt; B[Preprocessing];
    B --&gt; C[Alignment &lt;br&gt; LD Annotation];
    C --&gt; D[GMM];
    D --&gt; E[Visualization];
    D --&gt; F[Colocalization];</code></pre>
<h2 id="preprocessing">Preprocessing</h2>
<p>Data sources and software used in our study (and supported by default):</p>
<p>Data Sources:</p>
<ul>
<li><strong>eQTLCatalogue</strong>: <a href="https://github.com/eQTL-Catalogue/eQTL-Catalogue-resources/blob/master/tabix/tabix_ftp_paths.tsv">Tabix Index</a></li>
<li><strong>GTEx</strong>: <a href="https://console.cloud.google.com/storage/browser/gtex-resources/GTEx_Analysis_v8_QTLs/GTEx_Analysis_v8_EUR_eQTL_all_associations;tab=objects?inv=1&amp;invt=Ab037A&amp;prefix=&amp;forceOnObjectsSortingFiltering=true">Google Cloud</a> or <a href="https://www.gtexportal.org/home/downloads/adult-gtex/qtl">Portal</a></li>
<li><strong>eQTLGen</strong>: <a href="https://www.eqtlgen.org/phase1.html">Official Site</a></li>
<li><strong>BBJ</strong>: <a href="http://jenger.riken.jp/en/result">Official Site</a></li>
<li><strong>1000G</strong>: <a href="https://www.cog-genomics.org/plink/1.9/resources#phase1">Plink2 Resource</a></li>
<li><strong>PopCell (AFR)</strong>: <a href="https://doi.org/10.1038/s41586-023-06422-9">Nature 2023</a>. <em>Restricted Access</em> - <a href="https://dataset.owey.io/doi/10.48802/owey.e4qn-9190">Apply Here</a>.</li>
</ul>
<p>Software:</p>
<ul>
<li><strong>S-LDXR</strong>: <a href="https://github.com/huwenboshi/s-ldxr/tree/master">GitHub Algo</a></li>
<li><strong>Plink1.9</strong>: <a href="https://www.cog-genomics.org/plink/1.9/">Official Site</a></li>
<li><strong>Cibersortx</strong>: <a href="https://cibersortx.stanford.edu/">Official Site</a> used for cell type proportion estimation from GTEx data.</li>
<li><strong>COLOC</strong>: <a href="https://cran.r-project.org/web/packages/coloc/index.html">CRAN Package</a> used for colocalization analysis (optional).</li>
</ul>
<h3 id="format-data-by-chromosome">Format Data by Chromosome</h3>
<p>To optimize Python loading times, we split and format the data by chromosome.</p>
<div class="tabbed-set tabbed-alternate" data-tabs="1:4"><input checked="checked" id="__tabbed_1_1" name="__tabbed_1" type="radio" /><input id="__tabbed_1_2" name="__tabbed_1" type="radio" /><input id="__tabbed_1_3" name="__tabbed_1" type="radio" /><input id="__tabbed_1_4" name="__tabbed_1" type="radio" /><div class="tabbed-labels"><label for="__tabbed_1_1">GTEx</label><label for="__tabbed_1_2">BBJ</label><label for="__tabbed_1_3">eQTLCatalogue</label><label for="__tabbed_1_4">PopCell (AFR)</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<p>To process <strong>GTEx Whole Blood</strong> data:</p>
<p><code>shell/GTEx_Preprocess.sh</code></p>
<p><strong>Input Format</strong> (<code>GTEx_Analysis_v8_QTLs-GTEx_Analysis_v8_eQTL_all_associations-Whole_Blood.allpairs.txt.gz</code>)</p>
<table>
<thead>
<tr>
<th>gene_id</th>
<th>variant_id</th>
<th>tss_distance</th>
<th>ma_samples</th>
<th>ma_count</th>
<th>maf</th>
<th>pval_nominal</th>
<th>...</th>
</tr>
</thead>
<tbody>
<tr>
<td>ENSG00000227232.5</td>
<td>chr1_13550_G_A_b38</td>
<td>-16003</td>
<td>19</td>
<td>19</td>
<td>0.014</td>
<td>0.734</td>
<td>...</td>
</tr>
</tbody>
</table>
<p><strong>Output Format</strong> (<code>chr22.csv</code>)</p>
<table>
<thead>
<tr>
<th>GENE</th>
<th>RSID</th>
<th>CHR</th>
<th>POS</th>
<th>TSS_DISTANCE</th>
<th>A1</th>
<th>A2</th>
<th>MAF</th>
<th>PVAL</th>
<th>BETA</th>
<th>SE</th>
</tr>
</thead>
<tbody>
<tr>
<td>ENSG00000008735</td>
<td>rs117049661</td>
<td>22</td>
<td>49600902</td>
<td>-999783</td>
<td>T</td>
<td>C</td>
<td>0.0067</td>
<td>0.613</td>
<td>0.133</td>
<td>0.263</td>
</tr>
</tbody>
</table>
</div>
<div class="tabbed-block">
<p>To process <strong>BBJ cell type</strong> data:</p>
<p><code>shell/BBJ_Preprocess.sh</code></p>
<p><strong>Input Format</strong> (<code>chr22_cis_eqtl_mapping_nofilt_nomulti_with_alleles.txt.gz</code>)</p>
<table>
<thead>
<tr>
<th>SNP</th>
<th>POS</th>
<th>REF</th>
<th>ALT</th>
<th>gene</th>
<th>beta</th>
<th>t-stat</th>
<th>p-value</th>
</tr>
</thead>
<tbody>
<tr>
<td>chr22:16201313:I</td>
<td>16201313</td>
<td>A</td>
<td>AG</td>
<td>ENSG00000100181.17</td>
<td>0.114</td>
<td>0.355</td>
<td>0.722</td>
</tr>
</tbody>
</table>
<p><strong>Output Format</strong> (<code>chr22.csv</code>)</p>
<table>
<thead>
<tr>
<th>CHR</th>
<th>RSID</th>
<th>POS</th>
<th>A2</th>
<th>A1</th>
<th>GENE</th>
<th>BETA</th>
<th>Z</th>
<th>PVAL</th>
</tr>
</thead>
<tbody>
<tr>
<td>22</td>
<td>rs1000427</td>
<td>36890105</td>
<td>G</td>
<td>A</td>
<td>ENSG00000100055</td>
<td>-0.152</td>
<td>-0.858</td>
<td>0.392</td>
</tr>
</tbody>
</table>
</div>
<div class="tabbed-block">
<p>To process <strong>eQTLCatalogue</strong> data:</p>
<p><code>shell/eQTLCatalogue_Preprocess.sh</code></p>
<p><strong>Input Format</strong> (<code>QTD000031.all.tsv.gz</code>)</p>
<table>
<thead>
<tr>
<th>molecular_trait_id</th>
<th>chromosome</th>
<th>position</th>
<th>ref</th>
<th>alt</th>
<th>variant</th>
<th>...</th>
<th>beta</th>
<th>se</th>
</tr>
</thead>
<tbody>
<tr>
<td>ENSG00000187583</td>
<td>1</td>
<td>14464</td>
<td>A</td>
<td>T</td>
<td>chr1_14464_A_T</td>
<td>...</td>
<td>0.185</td>
<td>NP</td>
</tr>
</tbody>
</table>
<p><strong>Output Format</strong> (<code>chr22.csv</code>)</p>
<table>
<thead>
<tr>
<th>CHR</th>
<th>RSID</th>
<th>GENE</th>
<th>POS</th>
<th>A1</th>
<th>A2</th>
<th>BETA</th>
<th>SE</th>
<th>PVAL</th>
<th>Z</th>
<th>N</th>
</tr>
</thead>
<tbody>
<tr>
<td>22</td>
<td>rs5747203</td>
<td>ENSG00000015475</td>
<td>17493644</td>
<td>A</td>
<td>G</td>
<td>-0.040</td>
<td>0.148</td>
<td>0.786</td>
<td>-0.270</td>
<td>167</td>
</tr>
</tbody>
</table>
</div>
<div class="tabbed-block">
<p>To process <strong>African Population</strong> data from <em>Aquino et al. (2023)</em>:</p>
<p><strong>Source</strong>: <em>Dissecting human population variation in single-cell responses to SARS-CoV-2</em>, <strong>Nature</strong> (2023).</p>
<div class="admonition warning">
<p class="admonition-title">Data Access Control</p>
<p>This dataset requires specific application for access. 
Please visit the <strong><a href="https://dataset.owey.io/doi/10.48802/owey.e4qn-9190">Owey Dataset Portal</a></strong> for application details.</p>
</div>
<p><strong>Preprocessing Pipeline</strong>:</p>
<p>Please refer to the external repository <strong><a href="https://github.com/h-e-g/popCell_SARS-CoV-2">popCell_SARS-CoV-2</a></strong> for upstream processing scripts. 
Ensure the final output is formatted to match the <code>traceCB</code> standard (see other tabs).</p>
</div>
</div>
</div>
<h3 id="cell-type-information-proportion">Cell Type Information &amp; Proportion</h3>
<h4 id="eqtlcatalogue-id-and-cell-types">eQTLCatalogue ID and Cell Types</h4>
<table>
<thead>
<tr>
<th style="text-align: left;">ID</th>
<th style="text-align: left;">Dataset (N)</th>
<th style="text-align: left;">Cell Type</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: left;"><code>QTD000021</code></td>
<td style="text-align: left;">BLUEPRINT (191)</td>
<td style="text-align: left;">Monocytes</td>
</tr>
<tr>
<td style="text-align: left;"><code>QTD000069</code></td>
<td style="text-align: left;">CEDAR (286)</td>
<td style="text-align: left;">Monocytes</td>
</tr>
<tr>
<td style="text-align: left;"><code>QTD000081</code></td>
<td style="text-align: left;">Fairfax_2014 (420)</td>
<td style="text-align: left;">Monocytes</td>
</tr>
<tr>
<td style="text-align: left;"><code>QTD000031</code></td>
<td style="text-align: left;">BLUEPRINT (167)</td>
<td style="text-align: left;">CD4+ T cells</td>
</tr>
<tr>
<td style="text-align: left;"><code>QTD000067</code></td>
<td style="text-align: left;">CEDAR (290)</td>
<td style="text-align: left;">CD4+ T cells</td>
</tr>
<tr>
<td style="text-align: left;"><code>QTD000371</code></td>
<td style="text-align: left;">Kasela_2017 (280)</td>
<td style="text-align: left;">CD4+ T cells</td>
</tr>
<tr>
<td style="text-align: left;"><code>QTD000066</code></td>
<td style="text-align: left;">CEDAR (277)</td>
<td style="text-align: left;">CD8+ T cells</td>
</tr>
<tr>
<td style="text-align: left;"><code>QTD000372</code></td>
<td style="text-align: left;">Kasela_2017 (269)</td>
<td style="text-align: left;">CD8+ T cells</td>
</tr>
<tr>
<td style="text-align: left;"><code>QTD000073</code></td>
<td style="text-align: left;">CEDAR (262)</td>
<td style="text-align: left;">B cells</td>
</tr>
<tr>
<td style="text-align: left;"><code>QTD000115</code></td>
<td style="text-align: left;">Gilchrist_2021 (247)</td>
<td style="text-align: left;">NK cells</td>
</tr>
</tbody>
</table>
<h4 id="proportion-calculation">Proportion Calculation</h4>
<p>You can use any method to obtain cell type proportions. We recommend using GTEx whole blood TPM files with <strong>Cibersortx</strong>.</p>
<p><strong>Example Output (<code>cell_type_proportion.csv</code>)</strong>:
<div class="highlight"><pre><span></span><code>Cell_type,Proportion
B_cells,0.8069001744805046
PCs,0.30700693083630154
CD4+T_cells,10.401165801186464
...
</code></pre></div></p>
<h3 id="alignment">Alignment</h3>
<p>Run <code>src/preprocess/MergeChr.py</code> via <code>shell/run_merge.sh</code> to align all input data files. </p>
<ul>
<li>Results will be saved to <code>save_path_main</code>.</li>
<li>The cell type proportion file will strictly accompany the aligned data.</li>
</ul>
<p><strong>Directory Structure</strong>:
<div class="highlight"><pre><span></span><code>&lt;save_path_main&gt;/EAS_GTEx/QTD000021/
├── AUX_Monocytes/
├── INFO/
├── TAR_Monocytes/
└── Tissue/
</code></pre></div></p>
<h3 id="annotation-ld-scores">Annotation &amp; LD Scores</h3>
<h4 id="1-annotate-ld">1. Annotate LD</h4>
<p>Run <code>src/preprocess/create_ld_annot.py</code> via <code>shell/run_ld.sh</code>. This step prepares 1000G data for <code>s-ldxr</code>.</p>
<div class="admonition note dependency">
<p class="admonition-title">Note</p>
<p>Requires <code>pysnptools</code> and <code>statsmodels</code> to run s-ldxr. Ensure these are installed in your Python environment.</p>
</div>
<p><strong>Input</strong>: <code>1000G.&lt;pop&gt;.QC.maf.@.bed/bim/fam</code></p>
<p><strong>Output</strong>:
<div class="highlight"><pre><span></span><code>&lt;save_path_main&gt;/EAS_GTEx/QTD000021/LDSC/LD_annotation:
1.print_snps.txt
1.annot.gz
...
</code></pre></div></p>
<h4 id="2-run-s-ldxr">2. Run s-ldxr</h4>
<p>Use <code>shell/run_ld.sh</code> to calculate gene-level LD scores.</p>
<div class="admonition failure common error">
<p class="admonition-title">Failure</p>
<p><div class="highlight"><pre><span></span><code>IndexError: index 16464 is out of bounds for axis 0 with size 16464
</code></pre></div>
This usually indicates an alignment mismatch. <strong>Check if all input files are correctly aligned.</strong></p>
</div>
<h2 id="gmm-modeling">GMM Modeling</h2>
<p>After preprocessing, your study folder should follow this structure:</p>
<div class="highlight"><pre><span></span><code>.
├── AUX_Monocytes
├── celltype_proportion.csv
├── gene_snp_count
├── INFO
├── LDSC
├── TAR_Monocytes
└── Tissue
</code></pre></div>
<p>Execute <code>shell/run_gmm.sh</code> (wraps <code>src/traceCB/run_gmm.py</code>) to run the GMM model.</p>
<h3 id="output-files">Output Files</h3>
<p>The results are saved in Parquet notation for performance.</p>
<p><strong>1. Gene-level Results</strong> (<code>ENSG@.parquet</code>)</p>
<p>Contains detailed effect size estimates for Target (TAR), Auxiliary (AUX), and Tissue populations under different models (S: Summary, C: Cross-pop, T: Tissue-enhanced).</p>
<table>
<thead>
<tr>
<th>RSID</th>
<th>TAR_SBETA</th>
<th>TAR_CBETA</th>
<th>TAR_TBETA</th>
<th>TAR_SPVAL</th>
<th>...</th>
</tr>
</thead>
<tbody>
<tr>
<td>rs10985869</td>
<td>-0.513</td>
<td>-0.515</td>
<td>-0.637</td>
<td>0.0009</td>
<td>...</td>
</tr>
</tbody>
</table>
<p><strong>2. Chromosome Summary</strong> (<code>summary.csv</code>)</p>
<p>Contains heritability estimates ($h^2$) and effective sample sizes ($N_{eff}$).</p>
<table>
<thead>
<tr>
<th>GENE</th>
<th>NSNP</th>
<th>H1SQ</th>
<th>H2SQ</th>
<th>TAR_SNEFF</th>
<th>...</th>
</tr>
</thead>
<tbody>
<tr>
<td>ENSG...63</td>
<td>2321</td>
<td>7.49e-05</td>
<td>2.53e-04</td>
<td>269.05</td>
<td>...</td>
</tr>
</tbody>
</table>
<div class="admonition note performance optimization">
<p class="admonition-title">Note</p>
<p>We use <code>numba</code> with <code>jit</code> and <code>nogil</code> for high-performance computing. If you need to debug, you can comment out the <code>@</code> decorators in the source code, though this will significantly slow down execution.</p>
</div>
<h2 id="colocalization">Colocalization</h2>
<h3 id="prerequisites">Prerequisites</h3>
<ul>
<li><a href="https://cran.r-project.org/web/packages/LDlinkR/vignettes/LDlinkR.html">LDlinkR API Token</a> (Required for <code>src/coloc/run_ldlink.R</code>)</li>
<li><strong>Bedtools</strong>: <code>closestBed</code> binary</li>
<li><strong>References</strong>: hg19/GRCh37 cytoband, Gene Annotation BED</li>
</ul>
<h3 id="1-leading-snp-annotation">1. Leading SNP Annotation</h3>
<p>Run <code>src/coloc/find_leadingSNP.py</code>:</p>
<div class="highlight"><pre><span></span><code>python<span class="w"> </span>src/coloc/find_leadingSNP.py<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--gwas_sumstats_path<span class="w"> </span>&lt;PATH&gt;<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--save_path<span class="w"> </span>&lt;OUTPUT&gt;<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--save_prefix<span class="w"> </span>&lt;PREFIX&gt;
</code></pre></div>
<p><strong>Key Outputs:</strong>
*   <code>{prefix}_loci.csv</code>: Final result merging SNP positions with closest Ensembl gene IDs.</p>
<h3 id="2-run-coloc">2. Run COLOC</h3>
<p>Use <code>shell/run_coloc.sh</code> to execute the colocalization analysis for each study.</p>
<h2 id="visualization">Visualization</h2>
<p>Visualization scripts are located in <code>src/visual/</code>.</p>
<ul>
<li><strong>Logic</strong>: <code>src/visual/utils.py</code> selects which study to plot.</li>
<li><strong>Style</strong>: <code>src/visual/meta_data.json</code> defines colors, labels, and plot settings.</li>
</ul>
<p>To reproduce Figure 3 or Figure 5 from our paper, ensure the correct study is selected in <code>utils.py</code>.</p>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      
      <script id="__config" type="application/json">{"annotate": null, "base": "..", "features": ["navigation.tabs", "navigation.sections", "navigation.top", "search.suggest", "search.highlight", "content.code.copy"], "search": "../assets/javascripts/workers/search.2c215733.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../assets/javascripts/bundle.79ae519e.min.js"></script>
      
    
  </body>
</html>