
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="Trans-ancestry cell-type-specific eQTL effects mapping">
      
      
        <meta name="author" content="Luca Jiang">
      
      
        <link rel="canonical" href="https://lucajiang.github.io/traceCB/api_reference/">
      
      
        <link rel="prev" href="../pipeline/">
      
      
      
        
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.7.2">
    
    
      
        <title>API Reference - traceCB</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.484c7ddc.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.ab4e12ef.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#api-reference" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="traceCB" class="md-header__button md-logo" aria-label="traceCB" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            traceCB
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              API Reference
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="slate" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12s-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/lucajiang/traceCB" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    lucajiang/traceCB
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
            
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href=".." class="md-tabs__link">
        
  
  
    
  
  Home

      </a>
    </li>
  

      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../tutorial/" class="md-tabs__link">
        
  
  
    
  
  Tutorial

      </a>
    </li>
  

      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../pipeline/" class="md-tabs__link">
        
  
  
    
  
  Pipeline

      </a>
    </li>
  

      
        
  
  
  
    
  
  
    <li class="md-tabs__item md-tabs__item--active">
      <a href="./" class="md-tabs__link">
        
  
  
    
  
  API Reference

      </a>
    </li>
  

      
    </ul>
  </div>
</nav>
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="traceCB" class="md-nav__button md-logo" aria-label="traceCB" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    traceCB
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/lucajiang/traceCB" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    lucajiang/traceCB
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Home
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../tutorial/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Tutorial
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../pipeline/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Pipeline
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    
  
    API Reference
  

    
  </span>
  
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    
  
    API Reference
  

    
  </span>
  
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#tracecbgmm" class="md-nav__link">
    <span class="md-ellipsis">
      
        traceCB.gmm
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="traceCB.gmm">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#gmm" class="md-nav__link">
    <span class="md-ellipsis">
      
        GMM
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gmmtissue" class="md-nav__link">
    <span class="md-ellipsis">
      
        GMMtissue
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tracecbldsc" class="md-nav__link">
    <span class="md-ellipsis">
      
        traceCB.ldsc
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="traceCB.ldsc">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#run_single_ldsc" class="md-nav__link">
    <span class="md-ellipsis">
      
        Run_Single_LDSC
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#run_cross_ldsc" class="md-nav__link">
    <span class="md-ellipsis">
      
        Run_Cross_LDSC
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#tracecbgmm" class="md-nav__link">
    <span class="md-ellipsis">
      
        traceCB.gmm
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="traceCB.gmm">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#gmm" class="md-nav__link">
    <span class="md-ellipsis">
      
        GMM
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gmmtissue" class="md-nav__link">
    <span class="md-ellipsis">
      
        GMMtissue
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tracecbldsc" class="md-nav__link">
    <span class="md-ellipsis">
      
        traceCB.ldsc
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="traceCB.ldsc">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#run_single_ldsc" class="md-nav__link">
    <span class="md-ellipsis">
      
        Run_Single_LDSC
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#run_cross_ldsc" class="md-nav__link">
    <span class="md-ellipsis">
      
        Run_Cross_LDSC
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              
              <article class="md-content__inner md-typeset">
                
                  


  
  


<h1 id="api-reference">API Reference</h1>
<p>This page provides detailed documentation for the core functions in the <code>traceCB</code> package.</p>
<ul>
<li><a href="#api-reference">API Reference</a></li>
<li><a href="#tracecbgmm">traceCB.gmm</a><ul>
<li><a href="#gmm"><code>GMM</code></a></li>
<li><a href="#gmmtissue"><code>GMMtissue</code></a></li>
</ul>
</li>
<li><a href="#tracecbldsc">traceCB.ldsc</a><ul>
<li><a href="#run_single_ldsc"><code>Run_Single_LDSC</code></a></li>
<li><a href="#run_cross_ldsc"><code>Run_Cross_LDSC</code></a></li>
</ul>
</li>
</ul>
<h2 id="tracecbgmm">traceCB.gmm</h2>
<p>Core functions for the Generalized Method of Moments (GMM) estimation.</p>
<h3 id="gmm"><code>GMM</code></h3>
<p>Apply cross-population GMM (without tissue-specific information) to estimate effect sizes.</p>
<div class="highlight"><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">GMM</span><span class="p">(</span>
    <span class="n">Omega</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
    <span class="n">C</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
    <span class="n">beta1</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span>
    <span class="n">se1</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span>
    <span class="n">ld1</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span>
    <span class="n">beta2</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span>
    <span class="n">se2</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span>
    <span class="n">ld2</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span>
    <span class="n">ldx</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">float</span><span class="p">,</span> <span class="nb">float</span><span class="p">,</span> <span class="nb">float</span><span class="p">,</span> <span class="nb">float</span><span class="p">]</span>
</code></pre></div>
<p><strong>Parameters</strong></p>
<ul>
<li><strong>Omega</strong> (<code>np.ndarray</code>): A (2, 2) per-SNP covariance matrix.</li>
<li><strong>C</strong> (<code>np.ndarray</code>): A (2, 2) genetic drift matrix estimated from LDSC.</li>
<li><strong>beta1</strong> (<code>float</code>): Effect size (beta) for the SNP in population 1.</li>
<li><strong>se1</strong> (<code>float</code>): Standard error for the SNP in population 1.</li>
<li><strong>ld1</strong> (<code>float</code>): LD score between the SNP and the rest of the SNPs in the target gene in population 1.</li>
<li><strong>beta2</strong> (<code>float</code>): Effect size (beta) for the SNP in population 2.</li>
<li><strong>se2</strong> (<code>float</code>): Standard error for the SNP in population 2.</li>
<li><strong>ld2</strong> (<code>float</code>): LD score between the SNP and the rest of the SNPs in the target gene in population 2.</li>
<li><strong>ldx</strong> (<code>float</code>): Cross-population LD score for the SNP between population 1 and population 2.</li>
</ul>
<p><strong>Returns</strong></p>
<ul>
<li><strong>beta1_blue</strong> (<code>float</code>): GMM estimate (BLUE) for population 1.</li>
<li><strong>se1_blue</strong> (<code>float</code>): Standard error of the GMM estimate for population 1.</li>
<li><strong>beta2_blue</strong> (<code>float</code>): GMM estimate (BLUE) for population 2.</li>
<li><strong>se2_blue</strong> (<code>float</code>): Standard error of the GMM estimate for population 2.</li>
</ul>
<hr />
<h3 id="gmmtissue"><code>GMMtissue</code></h3>
<p>Apply cross-population GMM incorporating tissue-specific information (e.g., from scRNA-seq derived eQTLs).</p>
<div class="highlight"><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">GMMtissue</span><span class="p">(</span>
    <span class="n">Omega</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
    <span class="n">C</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
    <span class="n">beta1</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span>
    <span class="n">se1</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span>
    <span class="n">ld1</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span>
    <span class="n">beta2</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span>
    <span class="n">se2</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span>
    <span class="n">ld2</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span>
    <span class="n">ldx</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span>
    <span class="n">beta_t</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span>
    <span class="n">se_t</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span>
    <span class="n">pi2_omega_o</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span>
    <span class="n">propt</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">float</span><span class="p">,</span> <span class="nb">float</span><span class="p">,</span> <span class="nb">float</span><span class="p">,</span> <span class="nb">float</span><span class="p">]</span>
</code></pre></div>
<p><strong>Parameters</strong></p>
<ul>
<li><strong>Omega</strong> (<code>np.ndarray</code>): A (2, 2) per-SNP covariance matrix.</li>
<li><strong>C</strong> (<code>np.ndarray</code>): A (3, 3) genetic drift matrix estimated from LDSC.</li>
<li><strong>beta1</strong> (<code>float</code>): Effect size (beta) for the SNP in population 1.</li>
<li><strong>se1</strong> (<code>float</code>): Standard error for the SNP in population 1.</li>
<li><strong>ld1</strong> (<code>float</code>): LD score for population 1.</li>
<li><strong>beta2</strong> (<code>float</code>): Effect size (beta) for the SNP in population 2.</li>
<li><strong>se2</strong> (<code>float</code>): Standard error for the SNP in population 2.</li>
<li><strong>ld2</strong> (<code>float</code>): LD score for population 2.</li>
<li><strong>ldx</strong> (<code>float</code>): Cross-population LD score.</li>
<li><strong>beta_t</strong> (<code>float</code>): Effect size (beta) for the SNP in the specific tissue/cell-type.</li>
<li><strong>se_t</strong> (<code>float</code>): Standard error for the SNP in the specific tissue/cell-type.</li>
<li><strong>pi2_omega_o</strong> (<code>float</code>): Variance component of other cell types within the tissue.</li>
<li><strong>propt</strong> (<code>float</code>): Proportion of the specific cell type in the tissue.</li>
</ul>
<p><strong>Returns</strong></p>
<ul>
<li><strong>beta1_blue</strong> (<code>float</code>): GMM estimate for population 1.</li>
<li><strong>se1_blue</strong> (<code>float</code>): Standard error for population 1.</li>
<li><strong>beta2_blue</strong> (<code>float</code>): GMM estimate for population 2.</li>
<li><strong>se2_blue</strong> (<code>float</code>): Standard error for population 2.</li>
</ul>
<hr />
<h2 id="tracecbldsc">traceCB.ldsc</h2>
<p>Functions for running Single and Cross-Population LD Score Regression (LDSC).</p>
<h3 id="run_single_ldsc"><code>Run_Single_LDSC</code></h3>
<p>Run LDSC regression to estimate heritability ($h^2$) in a single population.</p>
<div class="highlight"><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">Run_Single_LDSC</span><span class="p">(</span>
    <span class="n">zscore</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
    <span class="n">n</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
    <span class="n">ldscore</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
    <span class="n">intercept</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">float</span><span class="p">,</span> <span class="nb">float</span><span class="p">]</span>
</code></pre></div>
<p><strong>Parameters</strong></p>
<ul>
<li><strong>zscore</strong> (<code>np.ndarray</code>): Array of Z-scores for SNPs, shape <code>(num_SNP,)</code>.</li>
<li><strong>n</strong> (<code>np.ndarray</code>): Array of sample sizes for each SNP, shape <code>(num_SNP,)</code>.</li>
<li><strong>ldscore</strong> (<code>np.ndarray</code>): Array of LD scores, shape <code>(num_SNP,)</code>.</li>
<li><strong>intercept</strong> (<code>float</code>, optional): Fixed intercept value. If <code>np.nan</code> (default), the intercept is estimated from the data.</li>
</ul>
<p><strong>Returns</strong></p>
<ul>
<li><strong>h2</strong> (<code>float</code>): Estimated heritability. (Clipped to be $\ge$ MIN_HERITABILITY).</li>
<li><strong>h2_se</strong> (<code>float</code>): Standard error of the heritability estimate.</li>
</ul>
<hr />
<h3 id="run_cross_ldsc"><code>Run_Cross_LDSC</code></h3>
<p>Run LDSC regression to estimate genetic covariance ($\Omega$) between two populations.</p>
<div class="highlight"><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">Run_Cross_LDSC</span><span class="p">(</span>
    <span class="n">zscore1</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
    <span class="n">n1</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
    <span class="n">ldscore1</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
    <span class="n">zscore2</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
    <span class="n">n2</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
    <span class="n">ldscore2</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
    <span class="n">crossld</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
    <span class="n">intercept</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">]),</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]</span>
</code></pre></div>
<p><strong>Parameters</strong></p>
<ul>
<li><strong>zscore1</strong> (<code>np.ndarray</code>): Z-scores for population 1.</li>
<li><strong>n1</strong> (<code>np.ndarray</code>): Sample sizes for population 1.</li>
<li><strong>ldscore1</strong> (<code>np.ndarray</code>): LD scores for population 1.</li>
<li><strong>zscore2</strong> (<code>np.ndarray</code>): Z-scores for population 2.</li>
<li><strong>n2</strong> (<code>np.ndarray</code>): Sample sizes for population 2.</li>
<li><strong>ldscore2</strong> (<code>np.ndarray</code>): LD scores for population 2.</li>
<li><strong>crossld</strong> (<code>np.ndarray</code>): Cross-population LD scores.</li>
<li><strong>intercept</strong> (<code>np.ndarray</code>, optional): Array of intercept values <code>[h11, h22, h12]</code>. Default is <code>[nan, nan, nan]</code>, which estimates all intercepts.</li>
</ul>
<p><strong>Returns</strong></p>
<ul>
<li><strong>Omega</strong> (<code>np.ndarray</code>): Estimated genetic covariance matrix of shape <code>(2, 2)</code>.</li>
<li><code>Omega[0, 0]</code>: Heritability pop 1</li>
<li><code>Omega[1, 1]</code>: Heritability pop 2</li>
<li><code>Omega[0, 1]</code> / <code>Omega[1, 0]</code>: Genetic covariance</li>
<li><strong>Omega_se</strong> (<code>np.ndarray</code>): Standard error matrix for <code>Omega</code>, shape <code>(2, 2)</code>.</li>
</ul>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      
      <script id="__config" type="application/json">{"annotate": null, "base": "..", "features": ["navigation.tabs", "navigation.sections", "navigation.top", "search.suggest", "search.highlight", "content.code.copy"], "search": "../assets/javascripts/workers/search.2c215733.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../assets/javascripts/bundle.79ae519e.min.js"></script>
      
    
  </body>
</html>